### PR TITLE
fix(fslib): don't destroy createReadStream in ZipFS before it finishes

### PR DESCRIPTION
**Pull request by merceyz**
**What's the problem this PR addresses?**

There is a race condition in the `createReadStream` implementation in `ZipFS`, if the file the stream is created for is large enough the stream will be destroyed before the write finishes.

Fixes https://github.com/yarnpkg/berry/issues/2004

**How did you fix it?**

Avoid calling destroy on successful reads and let the `PassThrough` `autoDestroy` itself, applied the same changes to `createWriteStream`.
I also made the read async to speed it up a bit

**Checklist**
- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).
- [x] I have set the packages that need to be released for my changes to be effective.
- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/yarnpkg-fslib/tests/ZipFS.test.ts [hunk 1]
```diff
@@ -509,7 +509,8 @@ describe(`ZipFS`, () => {
 
     // Consuming the iterator should cause the Dir instance to close
 
-    expect(() => iter.next()).rejects.toThrow(`Directory handle was closed`);
+    // FIXME: This assertion fails
+    // await expect(() => iter.next()).rejects.toThrow(`Directory handle was closed`);
```
Comment (merceyz) [inline packages/yarnpkg-fslib/tests/ZipFS.test.ts:513]: This failed before this PR but wasn't `await`ed so it wasn't caught

File: packages/yarnpkg-fslib/sources/ZipFS.ts [hunk 2]
```diff
@@ -439,42 +439,34 @@ export class ZipFS extends BasePortableFakeFS {
...
-        closeStream();
-        callback(error);
-      },
-    });
+    const fd = this.openSync(p, `r`);
+
+    const stream = Object.assign(
+      new PassThrough({
+        emitClose: true,
+        autoDestroy: true,
```
Comment (arcanis) [inline packages/yarnpkg-fslib/sources/ZipFS.ts:447]: Aren't both of those already [set to `true` by default](https://nodejs.org/api/stream.html#stream_new_stream_writable_options)?

File: packages/yarnpkg-fslib/sources/ZipFS.ts [hunk 3]
```diff
@@ -439,42 +439,34 @@ export class ZipFS extends BasePortableFakeFS {
...
+        destroy: (error, callback) => {
+          clearImmediate(immediate);
+          this.closeSync(fd);
+          callback(error);
+        },
+      }),
+      {
+        close() {
+          stream.destroy();
```
Comment (arcanis) [inline packages/yarnpkg-fslib/sources/ZipFS.ts:456]: Does it support calling `destroy` multiple times? The old code seemed to tolerate it (the fd was cleared, so we weren't calling `closeSync` multiple times), the new one not.

File: packages/yarnpkg-fslib/sources/ZipFS.ts (hunk 2 above)
Comment (merceyz) [reply packages/yarnpkg-fslib/sources/ZipFS.ts:447]: Depends on the node version, check the history for the docs you linked to

File: packages/yarnpkg-fslib/sources/ZipFS.ts (hunk 3 above)
Comment (merceyz) [reply packages/yarnpkg-fslib/sources/ZipFS.ts:456]: Yes, the stream handles that internally

**Review by arcanis — APPROVED** [review]

**Pull request merged**